File: perl/4.pl
#!/usr/bin/perl
use List::Util qw(max);
use strict;

# Find the largest palindrome made from the product of two 3-digit numbers.

sub get_racecars{
  my ($i, $max, @racecars) = @_;
  return @racecars if $i > $max;
  return get_racecars(++$i, $max, (@racecars, grep { $_ eq reverse($_) } map { $i * $_ } $i..$max));
}

print max get_racecars(900, 999);
print "\n";

File: haskell/4.hs
-- Find the largest palindrome made from the product of two 3-digit numbers.

main = do
    print $ maximum $ filter isPalindrome (numberList 900 999)

numberList :: Integer -> Integer -> [Integer]
numberList start finish = do 
    foldr (++) [] $ map (\x -> map (*x) [start..finish]) [start..finish]

isPalindrome :: Integer -> Bool
isPalindrome n = if (show n) == reverse (show n) then True else False
